[SUP-808] Improve performance of WorkspaceSelector with many workspaces

nawatts opened the pull request:
The WorkspaceSelector component uses [react-select](https://www.npmjs.com/package/react-select) under the hood. By default, that library renders every option. For users that have hundreds or thousands of workspaces, this results in very poor performance when trying to select anything from that list. To improve performance, this virtualizes the list of options (using [react-virtualized](https://www.npmjs.com/package/react-virtualized)) This way, only the visible options are rendered.

Tested this by modifying `useWorkspaces` to create a list of 1000 workspaces and store it in `workspacesStore`.

## Before

Despite what it looks like, I started typing immediately after the menu opens. The menu takes that long to update.

https://user-images.githubusercontent.com/1156625/220777966-460902c4-e1b8-474c-b27f-57ef47b91de9.mov

## After

https://user-images.githubusercontent.com/1156625/220777954-098e63a2-b202-41ca-96b6-6c1956c36355.mov

nawatts committed:
Improve performance of WorkspaceSelector with many workspaces

nawatts committed:
Update select helper in e2e tests

cahrens commented:
@nawatts this is awesome-- this has been very painful when there are lots of workspaces, as well as when there are lots of billing projects (ie., that's a place were we could reuse this approach). I would love to review this and will try to get to it today.

nawatts commented:
Thanks @cahrens. Looking at this again, it does seem more generalizable than I had initially thought it would be. So maybe it should become another Select component instead of being tied to the workspaces menu.

cahrens commented:
> Thanks @cahrens. Looking at this again, it does seem more generalizable than I had initially thought it would be. So maybe it should become another Select component instead of being tied to the workspaces menu.

That would be awesome, because we had a user (a reseller) who was unable to create a Billing Project because they had so many Google accounts accessible to their email address that they struggled to ever load and select a Google account in the Create Terra Billing project dialog. The problem is discussed in the description of https://broadworkbench.atlassian.net/browse/WOR-409. We improved this by filtering out accounts that don't have firecloud permissions, but the fundamental performance issue was not addressed.

cahrens commented:
@nawatts I tried this out and found the following pretty confusing-- when I first triggered the dropdown, it said "No options". I closed it and tried again… once again, "No options". Then the third time options displayed (and this behavior was reproducible for me 3 times in a row). Can you display something like "Loading options" while it is in the initial state of the workspaces loading? Otherwise people are likely to give up and think that they have no workspaces they can use. And will it ever start displaying options without me having to close and reopen it?

nawatts commented:
I see the "No options" message while the workspaces are loading (though that's also the behavior of the current select). But I haven't been able to reproduce requiring the menu to be reopened to show options once they are loaded. I've been trying this out on the import workflow page. Which page are you seeing that on?

https://user-images.githubusercontent.com/1156625/221650073-d2d439d9-83df-4289-a252-95f46fcdd835.mov

cahrens commented:
@nawatts also import workflow. I probably just haven't waited long enough because I'm impatient… If I wait it does eventually render the options, but it takes 5 seconds or so. It would be awesome if it said something besides "No options" during that time.

I'm not seeing the issue that you were addressing with the hermione account (200ish workspaces). What account are you seeing it with? Or did you just hack the ajax response to have lots of workspaces?

nawatts commented:
> What account are you seeing it with? Or did you just hack the ajax response to have lots of workspaces?

I modified `useWorkspaces` to return 1000 workspaces.

nawatts committed:
Create VirtualizedSelect component

nawatts committed:
Show message while loading workspaces on import workflow page

nawatts commented:
- Rebased on dev to pull in some other changes to the import workflow page.
- Created a VirtualizedSelect component so that other menus can use this and had WorkspaceSelector use that.
- On the import workflow page, replaced "No options" with "Loading workspaces" while workspaces are loading.

nawatts commented:
The failing test has to do with react-virtualized's AutoSizer telling the List that it's 0 pixels wide.

nawatts committed:
Fix test

MatthewBemis reviewed (approved):
nice!

cahrens commented on src/components/common/Select.js:
```diff
@@ -130,3 +132,65 @@ export const AsyncCreatableSelect = props => {
+
+  // When navigating option with arrow keys, scroll the virtualized list so that the focused option is visible.
+  useEffect(() => {
+    const renderedOptions = Children.map(children, child => child.props.data)
```
Nit: could make a helper method out of lines 155-159 and use it for lines 145-149 also.

nawatts replied:
Done in 93925ed4f5e7da406b7bd10f952812b297f8d6f6.

cahrens commented on src/components/common/Select.js:
```diff
@@ -130,3 +132,65 @@ export const AsyncCreatableSelect = props => {
+  }
+
+  const rowCount = children.length
+  const rowHeight = 40
```
Why do we have to specify the row height vs. it being determined by the rendered list item?

nawatts replied:
react-virtualized requires it: https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#prop-types. I think so that it can handle scrolling within a virtualized list.

cahrens reviewed (approved):
I had one nit and a question, but otherwise LGTM. 

I will create a ticket on the workspaces board to reuse VirtualizedSelect on the billing project page. Thanks for pulling it out!

nawatts committed:
Add helper function to scroll to option for value
